### ISSUE TITLE
Caller needs to know move returned as `INVALID_MOVE`

### ISSUE THREAD

**Issue by flamecoals** [nicolodavis/boardgame.io#592]
It would be useful to also know the reason of the invalid move, so allow for some extra data specific to the game.

**Comment by delucis** [comment]
To continue the thread from Gitter. Couple of options/considerations:

### Moves could throw an error

```js
// define custom error class
class InvalidMoveError extends Error {
  constructor(data) {
    super('INVALID_MOVE');
    this.name = "InvalidMoveError";
    this.data = data;
  }
}
```
```js
// move definition
const { InvalidMoveError } = require('boardgame.io/core');

function move() {
  // move is invalid, throw an error
  throw new InvalidMoveError({ message: 'something went wrong', player: '2' });
}
```
```js
// process move
try {
  moveFn(G, ctx, ...args);
} catch (error) {
  if (!(error instanceof InvalidMoveError)) throw error;
  // return error.data to client
}
```

#### Notes

- It’s idiomatic Javascript — an invalid move is an error.


### Moves could return an error object

```js
// define error creator
function InvalidMove(data) {
  return { error: 'INVALID_MOVE', data }; // for example
}
```
```js
const { InvalidMove } = require('boardgame.io/core');

function move() {
  // move is invalid, return an error object
  return InvalidMove({ message: 'something went wrong', player: '2' });
}
```
```js
// process move
const result = moveFn(G, ctx, ...args);
if (result && result.error === 'INVALID_MOVE') {
  // move is invalid, return result.data the client
}
```

#### Notes

- This is a little closer to the current `return INVALID_MOVE` approach than throwing an error.


### `ctx` could have an error method

On the surface this seems pretty similar to an `InvalidMove` function to me, just that the function is attached to `ctx` instead of exported in core.

The major advantage of this would presumably be that it could hook into `ctx` in deeper ways — perhaps firing other kinds of changes to state or flow. What would those be?

**Comment by nicolodavis** [comment]
One way to work around the asynchronous issue is to return a Promise from the move dispatcher.  We just need to ensure that the client can still be called in a synchronous way when using an in-memory store.

**Comment by nicolodavis** [comment]
Of the approaches that we've discussed so far, I think I'm leaning toward moves returning an error object.

**Comment by jeetmehta** [comment]
👋 Hi there! My friends and I are using boardgame.io and the framework has been great. I'm currently trying to indicate to the user/player _why_ their move might be invalid.

Looks like this is an active stream of work? Is there an alternate recommendation or workaround that we can take in the meantime?

I've tried setting things within G (i.e. `G.invalidMove` and `G.invalidMoveErrorMessage`) but returning `INVALID_MOVE` just resets G to the prior state, so the client doesn't see those props change.

Would love any advice 🙏

**Comment by shaoster** [comment]
> 👋 Hi there! My friends and I are using boardgame.io and the framework has been great. I'm currently trying to indicate to the user/player _why_ their move might be invalid.
> 
> Looks like this is an active stream of work? Is there an alternate recommendation or workaround that we can take in the meantime?
> 
> I've tried setting things within G (i.e. `G.invalidMove` and `G.invalidMoveErrorMessage`) but returning `INVALID_MOVE` just resets G to the prior state, so the client doesn't see those props change.
> 
> Would love any advice 🙏 cc: @delucis @nicolodavis @shaoster

Hello! I'm currently working with @delucis to add a feature to help with this exact use case.

In the meantime, see https://github.com/boardgameio/boardgame.io/issues/292#issuecomment-431528906.

The prescriptive TL;DR there is to create a (hopefully pure) shared function (i.e. `isValidOrWhyNot`) that's called by both your UI and your move implementation.

For example: https://github.com/shaoster/unmuted2021/blob/3b4cb94f0828f84d146ff165b5ce211bedfd8039/src/Action.js#L63

**Comment by jeetmehta** [comment]
Hmm... the tricky part here is that we're trying to validate the _turn_ as opposed to the individual move. A single turn can have many moves. Apologies, I should've been clearer in my original comment.

> Sometimes you may need to go deep into the move (pathfinding, for example) before you realize that there isn't a valid outcome. In these cases, it might also be expensive to both validate and run the move every time in the cases where the move is actually valid.

This comment stands out to me since it's relevant to our game. We're building something similar to Scrabble, so there's a lot of word validation / traversal that happens before we can say that "the tiles you played this turn are invalid for x, y , z reasons".

Wondering if that changes or affects your recommendation?

**Comment by jeetmehta** [comment]
Scratch that - your recommendation worked! I just created a separate `validateTurn` function which is called before the turn is ended.

Thank you!! 🙏

**Comment by shaoster** [comment]
Yeah, in general you may want your validate function to additionally return some intermediate artifact (maybe a pruned search tree or list of valid moves) if the computation to get there is nontrivial.
